optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/BiAccumulator.java

```diff
@@ -46,28 +44,22 @@ public BiAccumulator(Variable<A> varA, Variable<B> varB,
     }
 
     @Override
-    public Object accumulate(Object workingMemoryContext, Object context, Tuple leftTuple, InternalFactHandle handle,
-            Declaration[] declarations, Declaration[] innerDeclarations, ReteEvaluator reteEvaluator) {
-        if (declarationA == null) {
-            init(leftTuple, innerDeclarations);
-        }
-
-        A a = extractValue(declarationA, offsetToA, leftTuple);
-        B b = extractValue(declarationB, offsetToB, leftTuple);
-        return accumulator.apply((ResultContainer_) context, a, b);
+    protected Runnable accumulate(ResultContainer_ context, Tuple leftTuple, InternalFactHandle handle,
+            Declaration[] innerDeclarations) {
+        A a = valueExtractorA.apply(leftTuple);
+        B b = valueExtractorB.apply(leftTuple);
+        return accumulator.apply(context, a, b);
     }
 
-    private void init(Tuple leftTuple, Declaration[] innerDeclarations) {
+    @Override
+    protected void initialize(Tuple leftTuple, Declaration[] innerDeclarations) {
```

Comment:
Why can't we do this code in the Accumulator.init(...) method that Drools guarantees to call once for us?
Can we do an RFE to @mariofusco for whatever guarantees is missing?

Reply:
I also explain that in the long comment.
I discussed this with Mario, and he considers `Accumulator` to be internal API - we are only using it to squeeze max performance out of Drools. As such, he doesn't want to do any changes to this.

Reply:
Let's create a DROOLS jira to spec out exactly what we want, and more importantly why.

If it's an internal API, then it doesn't need to be Backwards Compatible :)